### ISSUE TITLE
[HOLD #20908] [$1000] [mWeb] Redirects to Assign task instead of LHN on press back button of device

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#21356]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. click + icon
2. select Assign task option
3. enter title and click on next button
4. select Assignee and share somewhere
5. click on confirm task button to finish creating the task
6. press back button of device

## Expected Result: 
should navigate to LHN

## Actual Result: 
navigates to Assign task screen

## Workaround:
Can the user still use Expensify without this being fixed? Have you informed them of the workaround?

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [x] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [ ] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.26-1
**Reproducible in staging?:** Needs reproduction (For me it takes to LHN)
**Reproducible in production?:** Needs reproduction
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/43996225/349d536f-7333-4cbc-ac58-b4312344688f

**Expensify/Expensify Issue URL:**
**Issue reported by:** @mananjadhav 
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1686309062585369

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01ab0e974992d7d84a</li>
        <li>Upwork Job ID: 1672317686491295744</li>
        <li>Last Price Increase: 2023-07-07</li>
    </ul>
</details>

**Comment by joekaufmanexpensify** [comment]
[Clarifying](https://expensify.slack.com/archives/C049HHMV9SM/p1687470171492469?thread_ts=1686309062.585369&cid=C049HHMV9SM) if this bug is dev only

**Comment by gadhiyamanan** [comment]
@joekaufmanexpensify I think it is reproducible in staging and production but not sure because I have no access to task in staging or production. i tested on dev and reproducible in dev

**Comment by joekaufmanexpensify** [comment]
I can reproduce this. And then see on the native android app, clicking the device back button after you finish creating the task brings you back to LHN.

**Comment by ygshbht** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
The application navigates back to the "Assign task" screen when the back button of the device or browser is pressed, even after a task is successfully assigned. This occurs both on mobile and desktop platforms. The expected behavior in mobiles is for the application to navigate to the Left Hand Navigation (LHN) screen.

### What is the root cause of that problem?
This could be due to how the navigation stack or the webpage history is being managed. The "Assign task" screen might not be properly removed from the navigation history or stack after a task has been assigned, causing the application to navigate back to it when the device back button is pressed or the browser's back button is clicked.

### What changes do you think we should make in order to solve the problem?
A potential solution to this issue could be ensuring the "Assign task" screen is properly removed from the navigation history or stack once a task is successfully assigned. For web applications, manipulating the browser's history object can effectively achieve this – for instance, using the replaceState() method instead of pushState() can overwrite the current history entry instead of creating a new one. For a mobile application, invoking a navigation.popToTop() or a specific navigation.goBack() method might be effective.

**Comment by alitoshmatov** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
[mWeb] Redirects to Assign task instead of LHN on press back button of device 
### What is the root cause of that problem?
In the case of assign task there is two modals that is recorded in stack but when we create task we call `dismisModal` which removes one modal from stack history
https://github.com/Expensify/App/blob/116abf58876038cb102d94477c95ebb504a93c96/src/libs/actions/Task.js#L184
### What changes do you think we should make in order to solve the problem?
We can modify `dismissModal` to specify how many times it should pop the stack, like for this case we know that task creation is using two modals. We can pass second argument to `dismissModal` and based on this argument we can call 
```javascript
            navigationRef.current.dispatch({...StackActions.pop(), target: rootState.key});
``` 
multiple times.

### What alternative solutions did you explore? (Optional)

**Comment by s77rt** [comment]
@ygshbht Thanks for the proposal. Although your RCA _could_ be correct the writing is questionable, please re-write the RCA in clear statements describing the exact root cause instead of providing assumptions. Then please write a proper solution that actually reflects our code base.

**Comment by s77rt** [comment]
@alitoshmatov Thanks for the proposal. Can you please elaborate how we have two modals in this case?

BTW, we don't want to call `navigationRef.current.dispatch({...StackActions.pop(), target: rootState.key});` this will not navigate to the task report after being created.

**Comment by ygshbht** [comment]
@s77rt  I appreciate your feedback. Let me present a clearer version of the Root Cause Analysis and the Proposed Solution:

### Root Cause Analysis:
The application redirects to the "Assign task" screen when the back button is pressed, due to the **"Assign task" screen persisting in the navigation stack**.

### Proposed Solution:
Currently, the code is set up to remove only one screen (the "Confirm task" screen) from the navigation stack after task creation. However, we should adjust this to also remove the "Assign task" screen. To accomplish this, we can utilize a function named dismissNumModals, which can remove a specific number of screens from the navigation stack:

```
function dismissNumModals(num=1, targetReportID){
    const rootState = navigationRef.getRootState();

    Array(num).fill(0).forEach(repetition=>{
        navigationRef.current.dispatch({...StackActions.pop(), target: rootState.key});
    })

    if (targetReportID && targetReportID !== getTopmostReportId(rootState)) {
        const state = getStateFromPath(ROUTES.getReportRoute(targetReportID));

        const action = getActionFromState(state, linkingConfig.config);
        action.type = 'NAVIGATE';
        navigationRef.current.dispatch(action);
    }
}
```
By using this function, we will be able to remove both the "Confirm task" and "Assign task" screens, which should result in a proper redirect upon pressing the back button. I am confident that this solution will effectively address the issue. You are welcome to test this approach to verify its effectiveness.

![image](https://github.com/Expensify/App/assets/65986357/0252fa9b-2b29-485e-9182-f5bee93d628c)



> @ygshbht Thanks for the proposal. Although your RCA _could_ be correct the writing is questionable, please re-write the RCA in clear statements describing the exact root cause instead of providing assumptions. Then please write a proper solution that actually reflects our code base.

**Comment by s77rt** [comment]
@ygshbht Thanks, that's much better. Can you explain why this bug only occurs on Web (not reproducible on Native)?

**Comment by ygshbht** [comment]
@s77rt  Thank you for your query. 

The observed discrepancy in behavior stems from the divergent navigation mechanisms in web browsers and native apps.

In web browsers, activating the back button on the browser or invoking the back functionality propels the user to the preceding URL in the browser history stack. Consequently, as the "Assign task" screen has a corresponding state in the stack, pressing the back button will navigate back to this screen.

Contrastingly, native apps generally lack a universal "back" functionality akin to that in web browsers, excluding the back button present in the app's user interface or the system back button on Android devices. These buttons typically adhere to the application's navigation stack rather than the browser's history stack.

Similarly, when using the in-app UI to navigate backwards on the web app, the issue is not present, echoing the behavior of native apps. This again emphasizes that the issue lies in how the web browser handles the back action in contrast to the in-app UI or native apps.

Therefore, it's this variation in the navigation model that prevents the issue from manifesting on native platforms.

**Comment by alitoshmatov** [comment]
@s77rt 
I meant adding this to the begining of `dismissModal` and not change other parts, and it does navigate to report.
`
 for (let i = 0; i < additionalPopCount; i++) {
        navigationRef.current.dispatch({...StackActions.pop(), target: rootState.key});
    }
`

> Can you please elaborate how we have two modals in this case?

There are two modals, `New task` and `Confirm Task`

<img width="384" alt="Screenshot 2023-06-26 at 2 44 31 PM" src="https://github.com/Expensify/App/assets/59907218/820c7699-5b69-4edb-8a60-368cd088ebbc">

<img width="378" alt="Screenshot 2023-06-26 at 2 44 18 PM" src="https://github.com/Expensify/App/assets/59907218/eabf8795-d769-4afb-be99-c8391d294129">

**Result:**
https://github.com/Expensify/App/assets/59907218/9a10a3ed-c30f-4d01-91ff-fda4bb125559

This issue is also present in request and send money processes.

**Comment by s77rt** [comment]
@ygshbht Can we get more details where this difference is coming from? Is it related to `react-navigation`?

**Comment by s77rt** [comment]
@alitoshmatov Thanks. I got it

**Comment by ygshbht** [comment]
@s77rt  Thank you for your response.

In essence, the difference arises from how back navigation is handled in different environments. When we talk about native apps, the 'back' action typically uses the app's own navigation stack to determine where to go back to. The app directly manages this stack.

However, in a web browser, the back button on the toolbar or the system's back is used, which are managed differently than in-app navigation. The browser's back button navigates through the browser's history stack, not the app's navigation stack.

This is not related to the UI or react-navigation as the in-app navigation works consistently across web and native environments. It's the external back button on web browsers causing the difference.

I hope this helps to clarify. Let me know if there are still any areas of confusion!

**Comment by s77rt** [comment]
Moving discussion to slack https://expensify.slack.com/archives/C01GTK53T8Q/p1687794457811099

**Comment by s77rt** [comment]
@alitoshmatov, @ygshbht Can we remove the whole stack from history without the need to pass the number of screens? (would `popToTop` be any useful here?)

**Comment by ygshbht** [comment]
Yes, using **popToTop** from react-navigation would take us to the first screen in the stack, effectively removing all others. However, if you only want to remove certain screens (like "Assign task") and not the entire stack, then popToTop might not be suitable.

On the other hand, if we want to clear the entire stack and navigate to the home screen (or any other specific screen), using popToTop could indeed be a simple solution. It could be used like this:

```
navigationRef.current.dispatch(StackActions.popToTop());
```
[popToTop](https://reactnavigation.org/docs/stack-actions/#poptotop
)

**Comment by alitoshmatov** [comment]
I don't think it works, applied 
`navigationRef.current.dispatch({...StackActions.popToTop(), target: rootState.key});`
right before we navigate to report, and it is resulting in unexpected behavior:
https://github.com/Expensify/App/assets/59907218/2371471f-9c6d-4c57-af5f-8e24350d9fd9

@ygshbht 
Using it like this `navigationRef.current.dispatch(StackActions.popToTop());` didn't work for me

**Comment by ygshbht** [comment]
@alitoshmatov  Theoretically, it should work. However, if you're experiencing issues, it could be related to certain limitations in react-navigation or its interaction with the browser's history stack API. More exploration could also be beneficial to get to the heart of the matter. It's worth noting, though, that the preliminary solutions that have already been proposed tackle the problem effectively

**Comment by joekaufmanexpensify** [comment]
Still discussing next steps in Slack thread.

**Comment by s77rt** [comment]
@ygshbht, @alitoshmatov I think we are looking for better solutions here. Also please keep in touch with the Slack thread

**Comment by joekaufmanexpensify** [comment]
Still pending new proposals

**Comment by joekaufmanexpensify** [comment]
I don't think we need to adjust this quite yet. We've had a lot of interest here.

**Comment by joekaufmanexpensify** [comment]
Not overdue.

**Comment by s77rt** [comment]
@Nodebrute Thanks for the proposal. Your RCA about the stack is correct but I'm not sure about the reason you stated being the diff between native and web (`onNavigationMenuButtonClicked` is for the UI back button and not the native/hardware back button). The proposed solution will break the back functionality, we won't be able to go back to `NewTaskDetailsPage` once we land on `NewTaskPage`.

**Comment by joekaufmanexpensify** [comment]
Still discussing proposals here.

**Comment by joekaufmanexpensify** [comment]
@s77rt it sounds like we won't be able to move forward with @Nodebrute's solution, is that right? As the breaking the back button piece isn't a simple change the the proposal?

**Comment by s77rt** [comment]
@joekaufmanexpensify Right. We are not discussing proposals here. We are looking for new ones.

**Comment by s77rt** [comment]
@Nodebrute By resetting the state we will lose the history, won't we?

**Comment by s77rt** [comment]
@Nodebrute What I meant in that comment is to remove the modal stack screens only, e.g.
1. Open report A
2. Open report B
3. Open Assign Task modal
4. Continue in page 2 of Tasks and create task
5. Remove screens 3 and 4

The history of being in report A and B should persist.

**Comment by s77rt** [comment]
Melvin, not close, still looking for proposals

**Comment by joekaufmanexpensify** [comment]
Not overdue

**Comment by joekaufmanexpensify** [comment]
@s77rt Curious if you think this could be a good candidate to float to one of the agencies?

**Comment by s77rt** [comment]
@joekaufmanexpensify Sure! This is related to navigation. It would be probably faster to get someone who work on navigation related but this is not a necessity.

**Comment by joekaufmanexpensify** [comment]
[Reached out to SWM. ](https://expensify.slack.com/archives/C04878MDF34/p1689187790908689)

**Comment by WoLewicki** [comment]
It's the same as other issues related to how browser history works in connection to `react-navigation`. Even though we do a `replace` which should also replace the entries in browser stack, it only does it for the last entry, leaving all the e.g. `request money` flow screens still there, so clicking browser back button wants to navigate to those still. We are currently thinking of ways to manipulate the browser history stack in a correct way with @adamgrzybowski. We'll go back with updates on this ASAP. cc @mountiny can you add it to navigation rework follow-ups?

**Comment by mountiny** [comment]
@joekaufmanexpensify I went ahead and updated this PR to be on HOLD and weekly. the underlying RCA is same for multiple issues as @WoLewicki mentioned above and its not clear if there is a solution to this which would not break other stuff. 

SWM will investigate in coming weeks and adding this to the project will help us keep track of this

**Comment by joekaufmanexpensify** [comment]
Got it, sounds good. Thanks!

**Comment by joekaufmanexpensify** [comment]
Still on hold

**Comment by joekaufmanexpensify** [comment]
Same

**Comment by joekaufmanexpensify** [comment]
Same.

**Comment by joekaufmanexpensify** [comment]
Same

**Comment by joekaufmanexpensify** [comment]
Held

**Comment by mountiny** [comment]
@joekaufmanexpensify could you please confirm this issue is resolved in staging? If yes, then we can close it

**Comment by joekaufmanexpensify** [comment]
Sure, I'll test now.

**Comment by s77rt** [comment]
I don't think a checklist is needed here. The bug seems upstream (or at least it was introduced after navigation reboot). If anything we can go with whatever checklist is submitted here https://github.com/Expensify/App/issues/23796#issuecomment-1687085619 (the PR reviewer is assigned to that issue).

No payment is due here I think, @mananjadhav (bug reporter) can request from ND.

Nothing to do. :heavy_check_mark:

**Comment by gadhiyamanan** [comment]
I think this is eligible for reporting bonus because the PR was created after issue reported. The issue was reported by me not @mananjadhav 
Cc: @joekaufmanexpensify @mountiny

**Comment by joekaufmanexpensify** [comment]
My thinking was no reporting bonus would be issued here since my understanding of [this comment](https://github.com/Expensify/App/issues/21356#issuecomment-1634622203) was that the RCA of this issue was the same as other navigation initiatives we already planned to fix. LMK if that's not correct though @mountiny

**Comment by mountiny** [comment]
I agree with Joe here, basically this is not a unique case of the issue, but one of the symptoms/ one of many flows where you could run into it, creating money request, leaving a thread, creating a split, task all of those are same from fix perspective so it does not make sense to pay for each of them being reported. Sorry.
